### PR TITLE
fix: Fix Sniper printer failing to put added import statement on separate line

### LINKED ISSUE #3698: bug: Sniper fails to add newline when adding new import statement
For example, with this file:

```java
import java.util.Set;

public class Cls {}
```
adding an import statement will result in a pretty-print that looks something like this:

```java
import java.util.ArrayList;java.util.Set;

public class Cls {}
```

### PR DESCRIPTION
**Pull request by slarse**
Fix #3698 

This is ready for review, see comments further down :)

### CODE REVIEW

File: src/test/java/spoon/test/prettyprinter/TestSniperPrinter.java [hunk 1]
```diff
@@ -398,16 +418,7 @@ private void testSniper(String testClass, Consumer<CtType<?>> transformation, Bi
...
-			SniperJavaPrettyPrinter printer = new SniperJavaPrettyPrinter(launcher.getEnvironment());
-			printer.setPreprocessors(Collections.unmodifiableList(Arrays.<Processor<CtElement>>asList(
-					//remove unused imports first. Do not add new imports at time when conflicts are not resolved
-					new ImportCleaner().setCanAddImports(false),
-					//solve conflicts, the current imports are relevant too
-					new ImportConflictDetector(),
-					//compute final imports
-					new ImportCleaner()
-			)));
-			return printer;
```
Comment (slarse) [inline src/test/java/spoon/test/prettyprinter/TestSniperPrinter.java:410]: Removed as it interfered with my test cases (removed unused imports), and no other test cases actually rely on this functionality. Very unclear why it's here.

**Comment by monperrus** [comment]
LGTM, unwip?

**Comment by slarse** [comment]
@monperrus I thought this was not good enough a fix, and I was right: it doesn't fix the problem if you add an import statement to a project that has a package statement. The problem is an off-by-one error that seemingly only comes into play when adding stuff (any stuff) at the very top of the file.

Back to the drawing board, I need to actually fix the off-by-one error :)

**Comment by slarse** [comment]
New test case [shows the newline fail here](https://travis-ci.com/github/INRIA/spoon/jobs/442522194#L1255-L1273)

**Comment by slarse** [comment]
@monperrus I've _finally_ figured this out. The missing newline for an import following a package statement, and the missing newline between import statements, are really two separate problems. Anyway, here's the rundown.

### Two import statements ending up on the same line
This problem is occurs when adding a new import statement, which is printed first in the file, and being immediately followed by a pre-existing import statement. As the pre-existing import statement has index 0 in the fragment child list, any newline events created by the default printer will never be printed.

The fix here is to always print newline events when printing the first pre-existing import statement. I _think_ this might affect other child lists than import statements, but we can't unconditionally print newline events at the start of every child list as the default printer sometimes generates unwanted events (e.g. in if/else statements) that we definitely don't want. I think we will simply have to keep adding exceptions to this particular case as they pop up.

### A newly added import statement ending up on the same line as the package declaration
This occurs when adding a new import statement that is printed after the package declaration. Previously, the `visitCtPackageDeclaration` printed its own newline. This is a big no-no for compatibility with the sniper printer, as if it decides that an element is unaltered, it will print its original source code _without_ the newline. Therefore, in general, no single `CtElement`'s visit method must include a newline for the visited element; it must be a completely standalone printer event. Of course, composite elements may still insert newlines between their constituent elements, but it should not terminate with a newline, or stuff like this happens. I think we will see this problem crop up again.

The fix was simply to not let `visitCtPackageDeclaration()` print a terminating newline, and instead print it in `visitCtCompilationUnit()`. Something similar may be the problem in #3697.

Also see some additional comments in the code.

**Let me know if you want me to split this PR in two**, it's technically two independent fixes with similar symptoms.

File: src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java [hunk 2]
```diff
@@ -1084,14 +1084,20 @@ public void visitCtCompilationUnit(CtCompilationUnit compilationUnit) {
...
-				}
-				for (CtType<?> t : compilationUnit.getDeclaredTypes()) {
-					scan(t);
-				}
+			scan(compilationUnit.getPackageDeclaration());
+
+			CtPackage pkg = compilationUnit.getDeclaredPackage();
+			if (pkg != null && !pkg.isUnnamedPackage()) {
+				printer.writeln();
+			}
```
Comment (slarse) [inline src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java:1092]: This is the only difference here, the rest is just indentation. In combination with the fix in `visitCtPackageDeclaration`, this fixes newlines after package declarations in the sniper printer.

File: src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java [hunk 3]
```diff
@@ -1111,7 +1117,7 @@ public void visitCtPackageDeclaration(CtPackageDeclaration packageDeclaration) {
 		CtPackageReference ctPackage = packageDeclaration.getReference();
 		elementPrinterHelper.writeComment(ctPackage, CommentOffset.BEFORE);
 		if (!ctPackage.isUnnamedPackage()) {
-			elementPrinterHelper.writePackageLine(ctPackage.getQualifiedName());
+			elementPrinterHelper.writePackageStatement(ctPackage.getQualifiedName());
```
Comment (slarse) [inline src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java:1120]: This is the "fix in visitCtPackageDeclaration" I mention.

File: src/main/java/spoon/support/sniper/internal/AbstractSourceFragmentPrinter.java [hunk 4]
```diff
@@ -102,6 +102,11 @@ public int update(PrinterEvent event) {
 			//so skip printing of this comment
 			//comment will be printed at place where it belongs to - together with spaces
 			return -1;
+		} else if (event.getRole() == CtRole.DECLARED_IMPORT && fragmentIndex == 0) {
+			// this is the first pre-existing import statement, and so we must print all newline
+			// events unconditionally to avoid placing it on the same line as a newly added import
+			// statement. See PR #3702 for details
+			printStandardSpaces();
```
Comment (slarse) [inline src/main/java/spoon/support/sniper/internal/AbstractSourceFragmentPrinter.java:109]: This is the special handling of the first import statement. We may need to add more roles here as more problems crop up. Using only the condition `fragmentIndex == 0` causes unwanted newlines in if/else statements, and perhaps elsewhere as well.

File: src/main/java/spoon/reflect/visitor/ElementPrinterHelper.java [hunk 5]
```diff
@@ -348,9 +348,20 @@ public void writeImports(Collection<CtImport> imports) {
...
+		printer.writeln();
+	}
+
+	/**
+	 * Write a package statement.
+	 */
+	public void writePackageStatement(String packageQualifiedName) {
 		printer.writeKeyword("package").writeSpace();
-		writeQualifiedName(packageQualifiedName).writeSeparator(";").writeln();
+		writeQualifiedName(packageQualifiedName).writeSeparator(";");
```
Comment (slarse) [inline src/main/java/spoon/reflect/visitor/ElementPrinterHelper.java:364]: Here, I've added the method `writePackageStatement`, as I did not want to break the previous API by modifying `writePackageLine`.

**Comment by monperrus** [comment]
Excellent work and explanations, thanks a lot @slarse

**Comment by slarse** [comment]
You're welcome :). I learned a ton about the sniper printer doing this, so I think I'm ready to get back to tackling #3697 now.

Ps. it looks like merging my PR broke the build on master, but I'm pretty sure that wasn't my fault.

**Comment by monperrus** [comment]
> it looks like merging my PR broke the build on master, but I'm pretty sure that wasn't my fault.

yes, it's not your fault, it's a flaky check due to Github API Rate 
Limiting, we may remove this low value check.

**Pull request merged**